src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,479 @@
+  This definition represents the sequence $(A_n)$ in the proof as described above.
+
+With these definitions, we are able consider the case where the sequence terminates seperately,
+allowing us to prove `exists_subset_restrict_nonpos`.
+-/
+
+private def exists_one_div_lt (s : signed_measure α) (i j : set α) (n : ℕ) : Prop :=
+∃ k : set α, k ⊆ i \ j ∧ measurable_set k ∧ (1 / (n + 1) : ℝ) < s k
+
+private lemma exists_nat_one_div_lt_measure_of_not_negative (hi : ¬ s ≤[i \ j] 0) :
+  ∃ (n : ℕ), exists_one_div_lt s i j n :=
+let ⟨k, hj₁, hj₂, hj⟩ := exists_pos_measure_of_not_restrict_le_zero s hi in
+let ⟨n, hn⟩ := exists_nat_one_div_lt hj in ⟨n, k, hj₂, hj₁, hn⟩
+
+private def find_exists_one_div_lt (s : signed_measure α) (i j : set α) : ℕ :=
+if hi : ¬ s ≤[i \ j] 0 then nat.find (exists_nat_one_div_lt_measure_of_not_negative hi) else 0
+
+private lemma find_exists_one_div_lt_spec (hi : ¬ s ≤[i \ j] 0) :
+  exists_one_div_lt s i j (find_exists_one_div_lt s i j) :=
+begin
+  rw [find_exists_one_div_lt, dif_pos hi],
+  convert nat.find_spec _,
+end
+
+private lemma find_exists_one_div_lt_min (hi : ¬ s ≤[i \ j] 0) {m : ℕ}
+  (hm : m < find_exists_one_div_lt s i j) : ¬ exists_one_div_lt s i j m :=
+begin
+  rw [find_exists_one_div_lt, dif_pos hi] at hm,
+  exact nat.find_min _ hm
+end
+
+private def some_exists_one_div_lt (s : signed_measure α) (i j : set α) : set α :=
+if hi : ¬ s ≤[i \ j] 0 then classical.some (find_exists_one_div_lt_spec hi) else ∅
+
+private lemma some_exists_one_div_lt_spec (hi : ¬ s ≤[i \ j] 0) :
+  (some_exists_one_div_lt s i j) ⊆ i \ j ∧ measurable_set (some_exists_one_div_lt s i j) ∧
+  (1 / (find_exists_one_div_lt s i j + 1) : ℝ) < s (some_exists_one_div_lt s i j) :=
+begin
+  rw [some_exists_one_div_lt, dif_pos hi],
+  exact classical.some_spec (find_exists_one_div_lt_spec hi),
+end
+
+private lemma some_exists_one_div_lt_subset : some_exists_one_div_lt s i j ⊆ i \ j :=
+begin
+  by_cases hi : ¬ s ≤[i \ j] 0,
+  { exact let ⟨h, _⟩ := some_exists_one_div_lt_spec hi in h },
+  { rw [some_exists_one_div_lt, dif_neg hi],
+    exact set.empty_subset _ },
+end
+
+private lemma some_exists_one_div_lt_subset' : some_exists_one_div_lt s i j ⊆ i :=
+set.subset.trans some_exists_one_div_lt_subset (set.diff_subset _ _)
+
+private lemma some_exists_one_div_lt_measurable_set :
+  measurable_set (some_exists_one_div_lt s i j) :=
+begin
+  by_cases hi : ¬ s ≤[i \ j] 0,
+  { exact let ⟨_, h, _⟩ := some_exists_one_div_lt_spec hi in h },
+  { rw [some_exists_one_div_lt, dif_neg hi],
+    exact measurable_set.empty }
+end
+
+private lemma some_exists_one_div_lt_lt (hi : ¬ s ≤[i \ j] 0) :
+  (1 / (find_exists_one_div_lt s i j + 1) : ℝ) < s (some_exists_one_div_lt s i j) :=
+let ⟨_, _, h⟩ := some_exists_one_div_lt_spec hi in h
+
+private def restrict_nonpos_seq (s : signed_measure α) (i : set α) : ℕ → set α
+| 0 := some_exists_one_div_lt s i ∅
+| (n + 1) := some_exists_one_div_lt s i ⋃ k ≤ n,
```

Comment:
can you put parentheses around the second set in `some_exists_one_div_lt` to help the reader (Lean doesn't need them, sure!)